In [1]:
import pandas as pd

In [55]:
import re

In [2]:
df_tech_sp500=pd.read_csv("../../processed/tech_sp500_com_CIK_SIC_NAICS.csv")

In [3]:
df_tech_sp500

cik                           name entityType     sic  \
0     320193                     Apple Inc.  operating  3571.0   
1     789019                 MICROSOFT CORP  operating  7372.0   
2    1045810                    NVIDIA CORP  operating  3674.0   
3    1652044                  Alphabet Inc.  operating  7370.0   
4    1326801           Meta Platforms, Inc.  operating  7370.0   
..       ...                            ...        ...     ...   
119    78814          PITNEY BOWES INC /DE/  operating  3579.0   
120  1770450            Xerox Holdings Corp  operating  3577.0   
121    61986               MANITOWOC CO INC  operating  3531.0   
122  1881551       Solidion Technology Inc.  operating  3690.0   
123  1621672  Super League Enterprise, Inc.  operating  7373.0   

                                        sicDescription          ownerOrg  \
0                                 Electronic Computers     06 Technology   
1                        Services-Prepackaged Software     06 Technology   
2                     Semiconductors & Related Devices  04 Manufacturing   
3    Services-Computer Programming, Data Processing...     06 Technology   
4    Services-Computer Programming, Data Processing...     06 Technology   
..                                                 ...               ...   
119                               Office Machines, NEC     06 Technology   
120                 Computer Peripheral Equipment, NEC     06 Technology   
121                     Construction Machinery & Equip     06 Technology   
122  Miscellaneous Electrical Machinery, Equipment ...  04 Manufacturing   
123        Services-Computer Integrated Systems Design     06 Technology   

     insiderTransactionForOwnerExists  insiderTransactionForIssuerExists  \
0                                   0                                  1   
1                                   1                                  1   
2                                   1                                  1   
3                                   1                                  1   
4                                   0                                  1   
..                                ...                                ...   
119                                 0                                  1   
120                                 0                                  1   
121                                 1                                  1   
122                                 0                                  1   
123                                 0                                  1   

        tickers      exchanges  ...  mailing_stateOrCountry  mailing_zipCode  \
0          AAPL         Nasdaq  ...                      CA            95014   
1          MSFT         Nasdaq  ...                      WA       98052-6399   
2          NVDA         Nasdaq  ...                      CA            95051   
3    GOOGL,GOOG  Nasdaq,Nasdaq  ...                      CA            94043   
4          META         Nasdaq  ...                      CA            94025   
..          ...            ...  ...                     ...              ...   
119  PBI,PBI-PB      NYSE,NYSE  ...                      CT       06926-0700   
120         XRX         Nasdaq  ...                      CT            06851   
121         MTW           NYSE  ...                      WI            53224   
122         STI         Nasdaq  ...                      TX            75240   
123         SLE         Nasdaq  ...                      CA            90404   

     mailing_stateOrCountryDescription            business_street1  \
0                                   CA          ONE APPLE PARK WAY   
1                                   WA           ONE MICROSOFT WAY   
2                                   CA   2788 SAN TOMAS EXPRESSWAY   
3                                   CA   1600 AMPHITHEATRE PARKWAY   
4                                   CA                  1 META WAY   
..       

In [5]:
len(df_tech_sp500.name.unique())

124

So each company, if no company changed their name, should occur only once in the dataframe. We need to find a way to make sure there were no name changes, and the tickers should include all used tickers. 

In [6]:
list(df_tech_sp500.name.unique())

['Apple Inc.',
 'MICROSOFT CORP',
 'NVIDIA CORP',
 'Alphabet Inc.',
 'Meta Platforms, Inc.',
 'Broadcom Inc.',
 'ORACLE CORP',
 'Salesforce, Inc.',
 'ADOBE INC.',
 'ADVANCED MICRO DEVICES INC',
 'THERMO FISHER SCIENTIFIC INC.',
 'T-Mobile US, Inc.',
 'CISCO SYSTEMS, INC.',
 'DANAHER CORP /DE/',
 'INTUIT INC.',
 'TEXAS INSTRUMENTS INC',
 'INTERNATIONAL BUSINESS MACHINES CORP',
 'APPLIED MATERIALS INC /DE',
 'ServiceNow, Inc.',
 'VERIZON COMMUNICATIONS INC',
 'CATERPILLAR INC',
 'COMCAST CORP',
 'AT&T INC.',
 'MICRON TECHNOLOGY INC',
 'Eaton Corp plc',
 'LAM RESEARCH CORP',
 'Arista Networks, Inc.',
 'ANALOG DEVICES INC',
 'Palo Alto Networks Inc',
 'Medtronic plc',
 'AUTOMATIC DATA PROCESSING INC',
 'DEERE & CO',
 'INTEL CORP',
 'SYNOPSYS INC',
 'AMPHENOL CORP /DE/',
 'Dell Technologies Inc.',
 'CADENCE DESIGN SYSTEMS INC',
 'Trane Technologies plc',
 'Palantir Technologies Inc.',
 'ILLINOIS TOOL WORKS INC',
 'NXP Semiconductors N.V.',
 'CrowdStrike Holdings, Inc.',
 'CARRIER GLOBAL Cor

This is by chatgpt: 
Here is the list of companies with their previous names (if applicable):

    Apple Inc. - Previously known as Apple Computer, Inc.
    MICROSOFT CORP - No known prior name (changed only in legal structure).
    NVIDIA CORP - No known prior name (founded as NVIDIA Corporation).
    Alphabet Inc. - Previously known as Google Inc..
    Meta Platforms, Inc. - Previously known as Facebook, Inc..
    Broadcom Inc. - Previously Avago Technologies after acquiring Broadcom.
    ORACLE CORP - No prior name (founded as Oracle Corporation).
    Salesforce, Inc. - No known prior name (founded as Salesforce.com).
    ADOBE INC. - No prior name (founded as Adobe Systems).
    ADVANCED MICRO DEVICES INC - No known prior name (founded as AMD).
    THERMO FISHER SCIENTIFIC INC. - Previously known as Thermo Electron Corporation, merged with Fisher Scientific to form the current name.
    T-Mobile US, Inc. - Previously VoiceStream Wireless before rebranding to T-Mobile.
    CISCO SYSTEMS, INC. - No prior name (founded as Cisco Systems).
    DANAHER CORP /DE/ - No known prior name (founded as Danaher Corporation).
    INTUIT INC. - No prior name (founded as Intuit).
    TEXAS INSTRUMENTS INC - No prior name (founded as Texas Instruments).
    INTERNATIONAL BUSINESS MACHINES CORP - Previously Computing-Tabulating-Recording Company (CTR) before becoming IBM.
    APPLIED MATERIALS INC /DE - No prior name (founded as Applied Materials).
    ServiceNow, Inc. - No prior name (founded as ServiceNow).
    VERIZON COMMUNICATIONS INC - Formed from the merger of Bell Atlantic and GTE Corporation.
    CATERPILLAR INC - No prior name (founded as Caterpillar Tractor Co.).
    COMCAST CORP - Previously American Cable Systems, rebranded as Comcast.
    AT&T INC. - Previously American Telephone and Telegraph Company (AT&T), and underwent multiple mergers.
    MICRON TECHNOLOGY INC - No prior name (founded as Micron Technology).
    Eaton Corp plc - Previously Eaton Corporation before merging with Cooper Industries.
    LAM RESEARCH CORP - No prior name (founded as Lam Research).
    Arista Networks, Inc. - No prior name (founded as Arista Networks).
    ANALOG DEVICES INC - No prior name (founded as Analog Devices).
    Palo Alto Networks Inc - No prior name (founded as Palo Alto Networks).
    Medtronic plc - Previously known as Medtronic, Inc., restructured to a plc after merging with Covidien.
    AUTOMATIC DATA PROCESSING INC - No prior name (founded as Automatic Data Processing).
    DEERE & CO - No prior name (founded as Deere & Company).
    INTEL CORP - No prior name (founded as Intel).
    SYNOPSYS INC - No prior name (founded as Synopsys).
    AMPHENOL CORP /DE/ - No prior name (founded as Amphenol).
    Dell Technologies Inc. - Previously known as Dell Inc., changed after going private and later public again as Dell Technologies.
    CADENCE DESIGN SYSTEMS INC - No prior name (founded as Cadence Design Systems).
    Trane Technologies plc - Previously Ingersoll-Rand, rebranded after splitting into two entities.
    Palantir Technologies Inc. - No prior name (founded as Palantir Technologies).
    ILLINOIS TOOL WORKS INC - No prior name (founded as Illinois Tool Works).
    NXP Semiconductors N.V. - Previously Philips Semiconductors before being spun off and renamed to NXP.
    CrowdStrike Holdings, Inc. - No prior name (founded as CrowdStrike).
    CARRIER GLOBAL Corp - Previously part of United Technologies Corporation (UTC), spun off as Carrier.
    ROPER TECHNOLOGIES INC - Previously Roper Industries.
    Fortinet, Inc. - No prior name (founded as Fortinet).
    Autodesk, Inc. - No prior name (founded as Autodesk).
    CHARTER COMMUNICATIONS, INC. /MO/ - Previously Insight Communications, rebranded to Charter after merger.
    Johnson Controls International plc - Previously Johnson Controls Inc., changed after merging with Tyco International.
    MONOLITHIC POWER SYSTEMS INC - No prior name (founded as Monolithic Power Systems).
    MICROCHIP TECHNOLOGY INC - No prior name (founded as Microchip Technology).
    CUMMINS INC - No prior name (founded as Cummins Engine Company).
    ELECTRONIC ARTS INC. - No prior name (founded as Electronic Arts).
    Verisk Analytics, Inc. - No prior name (founded as Verisk Analytics).
    AMETEK INC/ - No prior name (founded as Ametek).
    COGNIZANT TECHNOLOGY SOLUTIONS CORP - No prior name (founded as Cognizant Technology Solutions).
    Ingersoll Rand Inc. - Previously Ingersoll Rand Company, changed after merging with Trane.
    Super Micro Computer, Inc. - No prior name (founded as Super Micro Computer).
    HP INC - Previously known as Hewlett-Packard Company, spun off HP Inc. and Hewlett Packard Enterprise.
    Baker Hughes Co - Previously Baker Hughes Incorporated, changed after merging with GE Oil & Gas.
    ON SEMICONDUCTOR CORP - Previously Motorola Semiconductor, renamed after spin-off.
    Xylem Inc. - Previously ITT Corporation, spun off into Xylem.
    ROCKWELL AUTOMATION, INC - Previously Rockwell International, rebranded to Rockwell Automation.
    ANSYS INC - No prior name (founded as Ansys).
    NetApp, Inc. - No prior name (founded as Network Appliance).
    Veralto Corp - No prior name (founded as Veralto).
    TAKE TWO INTERACTIVE SOFTWARE INC - Previously Take-Two Interactive, no significant name change.
    TYLER TECHNOLOGIES INC - No prior name (founded as Tyler Technologies).
    DOVER Corp - No prior name (founded as Dover Corporation).
    Hewlett Packard Enterprise Co - Spun off from Hewlett-Packard and rebranded.
    Fortive Corp - Previously Danaher Corporation, rebranded after spin-off.
    FIRST SOLAR, INC. - No prior name (founded as First Solar).
    Keysight Technologies, Inc. - Previously Agilent Technologies, spun off as Keysight.
    GoDaddy Inc. - No prior name (founded as GoDaddy).
    TERADYNE, INC - No prior name (founded as Teradyne).
    Seagate Technology Holdings plc - Previously known as Seagate Technology.
    WESTERN DIGITAL CORP - No prior name (founded as Western Digital).
    PTC INC. - Previously Parametric Technology Corporation.
    HUBBELL INC - No prior name (founded as Hubbell Incorporated).
    Leidos Holdings, Inc. - Previously SAIC Inc., rebranded after spin-off.
    Warner Bros. Discovery, Inc. - Previously WarnerMedia before merging with Discovery.
    ZEBRA TECHNOLOGIES CORP - No prior name (founded as Zebra Technologies).
    Fox Corp - Previously News Corporation, rebranded as Fox Corporation.
    VERISIGN INC/CA - No prior name (founded as Verisign).
    SKYWORKS SOLUTIONS, INC. - No prior name (founded as Skyworks Solutions).
    Enphase Energy, Inc. - No prior name (founded as Enphase Energy).
    FACTSET RESEARCH SYSTEMS INC - No prior name (founded as FactSet).
    Gen Digital Inc. - Previously Symantec Corporation, rebranded as Gen Digital.
    IDEX CORP /DE/ - No prior name (founded as IDEX Corporation).
    PENTAIR plc - Previously Pentair, Inc., changed after split.
    NORDSON CORP - No prior name (founded as Nordson Corporation).
    TRIMBLE INC. - No prior name (founded as Trimble Navigation Limited).
    JUNIPER NETWORKS INC - No prior name (founded as Juniper Networks).
    JACK HENRY & ASSOCIATES INC - No prior name (founded as Jack Henry & Associates).
    TechnipFMC plc - Previously Technip, merged with FMC Technologies.
    EPAM Systems, Inc. - No prior name (founded as EPAM Systems).
    F5, INC. - No prior name (founded as F5 Networks).
    ITT INC. - Previously ITT Corporation, rebranded to ITT.
    Qorvo, Inc. - Previously TriQuint Semiconductor, merged and renamed Qorvo.
    HASBRO, INC. - No prior name (founded as Hasbro).
    Paycom Software, Inc. - No prior name (founded as Paycom).
    Match Group, Inc. - No prior name (founded as Match Group).
    Dayforce, Inc. - No prior name (founded as Dayforce).
    Paramount Global - Previously known as ViacomCBS and CBS Corporation.
    CIENA CORP - No prior name (founded as Ciena Corporation).
    Altair Engineering Inc. - No prior name (founded as Altair Engineering).
    ACUITY BRANDS INC - No prior name (founded as Acuity Brands).
    SentinelOne, Inc. - No prior name (founded as SentinelOne).
    NOV Inc. - Previously National Oilwell Varco, rebranded to NOV.
    HashiCorp, Inc. - No prior name (founded as HashiCorp).
    FLOWSERVE CORP - No prior name (founded as Flowserve).
    Lumen Technologies, Inc. - Previously CenturyLink, rebranded as Lumen Technologies.
    DXC Technology Co - Previously Hewlett-Packard Enterprise (HP Enterprise) and Computer Sciences Corporation (CSC).
    DoubleVerify Holdings, Inc. - No prior name (founded as DoubleVerify).
    IPG PHOTONICS CORP - No prior name (founded as IPG Photonics).
    TERADATA CORP /DE/ - No prior name (founded as Teradata).
    TEGNA INC - Previously Gannett Company, rebranded as TEGNA.
    TripAdvisor, Inc. - No prior name (founded as TripAdvisor).
    SOLAREDGE TECHNOLOGIES, INC. - No prior name (founded as SolarEdge).
    DNOW Inc. - Previously DistributionNOW, formerly part of National Oilwell Varco.
    PITNEY BOWES INC /DE/ - No prior name (founded as Pitney Bowes).
    Xerox Holdings Corp - Previously Xerox Corporation.
    MANITOWOC CO INC - Previously The Manitowoc Company, rebranded as Manitowoc.
    Solidion Technology Inc. - No prior name (founded as Solidion).
    Super League Enterprise, Inc. - No prior name (founded as Super League Gaming).

By chatgpt again, format is [company name, current ticker, old tickers]: 

In [7]:
lst_gpt_gen_new_old_tic = [
    ["Apple Inc.", "AAPL", "None"],
    ["MICROSOFT CORP", "MSFT", "None"],
    ["NVIDIA CORP", "NVDA", "None"],
    ["Alphabet Inc.", "GOOGL/GOOG", "None"],
    ["Meta Platforms, Inc.", "META", "FB"],
    ["Broadcom Inc.", "AVGO", "BRCM"],
    ["ORACLE CORP", "ORCL", "None"],
    ["Salesforce, Inc.", "CRM", "None"],
    ["ADOBE INC.", "ADBE", "None"],
    ["ADVANCED MICRO DEVICES INC", "AMD", "None"],
    ["THERMO FISHER SCIENTIFIC INC.", "TMO", "None"],
    ["T-Mobile US, Inc.", "TMUS", "None"],
    ["CISCO SYSTEMS, INC.", "CSCO", "None"],
    ["DANAHER CORP /DE/", "DHR", "None"],
    ["INTUIT INC.", "INTU", "None"],
    ["TEXAS INSTRUMENTS INC", "TXN", "None"],
    ["INTERNATIONAL BUSINESS MACHINES CORP", "IBM", "None"],
    ["APPLIED MATERIALS INC /DE", "AMAT", "None"],
    ["ServiceNow, Inc.", "NOW", "None"],
    ["VERIZON COMMUNICATIONS INC", "VZ", "None"],
    ["CATERPILLAR INC", "CAT", "None"],
    ["COMCAST CORP", "CMCSA", "None"],
    ["AT&T INC.", "T", "None"],
    ["MICRON TECHNOLOGY INC", "MU", "None"],
    ["Eaton Corp plc", "ETN", "None"],
    ["LAM RESEARCH CORP", "LRCX", "None"],
    ["Arista Networks, Inc.", "ANET", "None"],
    ["ANALOG DEVICES INC", "ADI", "None"],
    ["Palo Alto Networks Inc", "PANW", "None"],
    ["Medtronic plc", "MDT", "None"],
    ["AUTOMATIC DATA PROCESSING INC", "ADP", "None"],
    ["DEERE & CO", "DE", "None"],
    ["INTEL CORP", "INTC", "None"],
    ["SYNOPSYS INC", "SNPS", "None"],
    ["AMPHENOL CORP /DE/", "APH", "None"],
    ["Dell Technologies Inc.", "DELL", "None"],
    ["CADENCE DESIGN SYSTEMS INC", "CDNS", "None"],
    ["Trane Technologies plc", "TT", "None"],
    ["Palantir Technologies Inc.", "PLTR", "None"],
    ["ILLINOIS TOOL WORKS INC", "ITW", "None"],
    ["NXP Semiconductors N.V.", "NXPI", "None"],
    ["CrowdStrike Holdings, Inc.", "CRWD", "None"],
    ["CARRIER GLOBAL Corp", "CARR", "None"],
    ["ROPER TECHNOLOGIES INC", "ROP", "None"],
    ["Fortinet, Inc.", "FTNT", "None"],
    ["Autodesk, Inc.", "ADSK", "None"],
    ["CHARTER COMMUNICATIONS, INC. /MO/", "CHTR", "None"],
    ["Johnson Controls International plc", "JCI", "None"],
    ["MONOLITHIC POWER SYSTEMS INC", "MPWR", "None"],
    ["MICROCHIP TECHNOLOGY INC", "MCHP", "None"],
    ["CUMMINS INC", "CMI", "None"],
    ["ELECTRONIC ARTS INC.", "EA", "None"],
    ["Verisk Analytics, Inc.", "VRSK", "None"],
    ["AMETEK INC/", "AME", "None"],
    ["COGNIZANT TECHNOLOGY SOLUTIONS CORP", "CTSH", "None"],
    ["Ingersoll Rand Inc.", "IR", "None"],
    ["Super Micro Computer, Inc.", "SMCI", "None"],
    ["HP INC", "HPQ", "None"],
    ["Baker Hughes Co", "BKR", "None"],
    ["ON SEMICONDUCTOR CORP", "ON", "None"],
    ["Xylem Inc.", "XYL", "None"],
    ["ROCKWELL AUTOMATION, INC", "ROK", "None"],
    ["ANSYS INC", "ANSS", "None"],
    ["NetApp, Inc.", "NTAP", "None"],
    ["Veralto Corp", "VRTX", "None"],
    ["TAKE TWO INTERACTIVE SOFTWARE INC", "TTWO", "None"],
    ["TYLER TECHNOLOGIES INC", "TYL", "None"],
    ["DOVER Corp", "DOV", "None"],
    ["Hewlett Packard Enterprise Co", "HPE", "None"],
    ["Fortive Corp", "FTV", "None"],
    ["FIRST SOLAR, INC.", "FSLR", "None"],
    ["Keysight Technologies, Inc.", "KEYS", "None"],
    ["GoDaddy Inc.", "GDDY", "None"],
    ["TERADYNE, INC", "TER", "None"],
    ["Seagate Technology Holdings plc", "STX", "None"],
    ["WESTERN DIGITAL CORP", "WDC", "None"],
    ["PTC INC.", "PTC", "None"],
    ["HUBBELL INC", "HUBB", "None"],
    ["Leidos Holdings, Inc.", "LDOS", "None"],
    ["Warner Bros. Discovery, Inc.", "WBD", "None"],
    ["ZEBRA TECHNOLOGIES CORP", "ZBRA", "None"],
    ["Fox Corp", "FOXA/FOXA", "None"],
    ["VERISIGN INC/CA", "VRSN", "None"],
    ["SKYWORKS SOLUTIONS, INC.", "SWKS", "None"],
    ["Enphase Energy, Inc.", "ENPH", "None"],
    ["FACTSET RESEARCH SYSTEMS INC", "FDS", "None"],
    ["Gen Digital Inc.", "GEN", "None"],
    ["IDEX CORP /DE/", "IEX", "None"],
    ["PENTAIR plc", "PNR", "None"],
    ["NORDSON CORP", "NDSN", "None"],
    ["TRIMBLE INC.", "TRMB", "None"],
    ["JUNIPER NETWORKS INC", "JNPR", "None"],
    ["JACK HENRY & ASSOCIATES INC", "JKHY", "None"],
    ["TechnipFMC plc", "FTI", "None"],
    ["EPAM Systems, Inc.", "EPAM", "None"],
    ["F5, INC.", "FFIV", "None"],
    ["ITT INC.", "ITT", "None"],
    ["Qorvo, Inc.", "QRVO", "None"],
    ["HASBRO, INC.", "HAS", "None"],
    ["Paycom Software, Inc.", "PAYC", "None"],
    ["Match Group, Inc.", "MTCH", "None"],
    ["Dayforce, Inc.", "None", "None"],
    ["Paramount Global", "PARA", "None"],
    ["CIENA CORP", "CIEN", "None"],
    ["Altair Engineering Inc.", "ALTR", "None"],
    ["ACUITY BRANDS INC", "AYI", "None"],
    ["SentinelOne, Inc.", "S", "None"],
    ["NOV Inc.", "NOV", "None"],
    ["HashiCorp, Inc.", "HCP", "None"],
    ["FLOWSERVE CORP", "FLS", "None"],
    ["Lumen Technologies, Inc.", "LUMN", "None"],
    ["DXC Technology Co", "DXC", "None"],
    ["DoubleVerify Holdings, Inc.", "DV", "None"],
    ["IPG PHOTONICS CORP", "IPGP", "None"],
    ["TERADATA CORP /DE/", "TDC", "None"],
    ["TEGNA INC", "TGNA", "None"],
    ["TripAdvisor, Inc.", "TRIP", "None"],
    ["SOLAREDGE TECHNOLOGIES, INC.", "SEDG", "None"],
    ["DNOW Inc.", "DNOW", "None"],
    ["PITNEY BOWES INC /DE/", "PBI", "None"],
    ["Xerox Holdings Corp", "XRX", "None"],
    ["MANITOWOC CO INC", "MTW", "None"],
    ["Solidion Technology Inc.", "None", "None"],
    ["Super League Enterprise, Inc.", "None", "None"]
]

Let's make sure that at least all companies in above list have a current ticker, we will correct it otherwise. 

In [27]:
for lst_ele in lst_gpt_gen_new_old_tic: 
    if lst_ele[1]=="None": 
        print(lst_ele, lst_gpt_gen_new_old_tic.index(lst_ele))

['Dayforce, Inc.', 'None', 'None'] 101
['Solidion Technology Inc.', 'None', 'None'] 122
['Super League Enterprise, Inc.', 'None', 'None'] 123


In [29]:
df_tech_sp500[df_tech_sp500.name=="Dayforce, Inc."].tickers.unique()[0]

'DAY'

In [30]:
for lst_ele in lst_gpt_gen_new_old_tic: 
    if lst_ele[1]=="None":
        lst_gpt_gen_new_old_tic[lst_gpt_gen_new_old_tic.index(lst_ele)][1]=df_tech_sp500[df_tech_sp500.name==lst_ele[0]].tickers.unique()[0]

In [32]:
for lst_ele in lst_gpt_gen_new_old_tic: 
    if lst_ele[1]=="None": 
        print(lst_ele, lst_gpt_gen_new_old_tic.index(lst_ele))
        
print("no ono")

no ono


Now that we have fixed the list, let's print out the ones that have previous tickers: 

In [35]:
for lst_ele in lst_gpt_gen_new_old_tic: 
    if lst_ele[2]!="None": 
        print(lst_ele, lst_gpt_gen_new_old_tic.index(lst_ele))

['Meta Platforms, Inc.', 'META', 'FB'] 4
['Broadcom Inc.', 'AVGO', 'BRCM'] 5


Let's update the dataframe on these bad apples. 

In [40]:
df_tech_sp500.at[4,"tickers"]=df_tech_sp500.at[4,"tickers"]+",FB"

In [42]:
df_tech_sp500.at[5,"tickers"]=df_tech_sp500.at[5,"tickers"]+",BRCM"

Just to double check the tickers in lst_gpt_gen_new_old_tic are all in the dataframe: 

In [46]:
lst_tech_sp500_tic=df_tech_sp500.tickers.to_list()

In [47]:
lst_tech_sp500_tic_sing=[]

for str_ele in lst_tech_sp500_tic: 
    lst_tech_sp500_tic_sing=lst_tech_sp500_tic_sing+str_ele.split(",")
    
lst_tech_sp500_tic_sing

['AAPL',
 'MSFT',
 'NVDA',
 'GOOGL',
 'GOOG',
 'META',
 'FB',
 'AVGO',
 'BRCM',
 'ORCL',
 'CRM',
 'ADBE',
 'AMD',
 'TMO',
 'TMUS',
 'CSCO',
 'DHR',
 'INTU',
 'TXN',
 'IBM',
 'AMAT',
 'NOW',
 'VZ',
 'CAT',
 'CMCSA',
 'CCZ',
 'T',
 'TBB',
 'TBC',
 'T-PA',
 'T-PC',
 'MU',
 'ETN',
 'LRCX',
 'ANET',
 'ADI',
 'PANW',
 'MDT',
 'ADP',
 'DE',
 'INTC',
 'SNPS',
 'APH',
 'DELL',
 'CDNS',
 'TT',
 'PLTR',
 'ITW',
 'NXPI',
 'CRWD',
 'CARR',
 'ROP',
 'FTNT',
 'ADSK',
 'CHTR',
 'JCI',
 'MPWR',
 'MCHP',
 'CMI',
 'EA',
 'VRSK',
 'AME',
 'CTSH',
 'IR',
 'SMCI',
 'HPQ',
 'BKR',
 'ON',
 'XYL',
 'ROK',
 'ANSS',
 'NTAP',
 'VLTO',
 'TTWO',
 'TYL',
 'DOV',
 'HPE',
 'HPE-PC',
 'FTV',
 'FSLR',
 'KEYS',
 'GDDY',
 'TER',
 'STX',
 'WDC',
 'PTC',
 'HUBB',
 'LDOS',
 'WBD',
 'ZBRA',
 'FOXA',
 'FOX',
 'VRSN',
 'SWKS',
 'ENPH',
 'FDS',
 'GEN',
 'IEX',
 'PNR',
 'NDSN',
 'TRMB',
 'JNPR',
 'JKHY',
 'FTI',
 'EPAM',
 'FFIV',
 'ITT',
 'QRVO',
 'HAS',
 'PAYC',
 'MTCH',
 'DAY',
 'PARA',
 'PARAA',
 'CIEN',
 'ALTR',
 'AYI',
 'S',

In [49]:
for lst_ele in lst_gpt_gen_new_old_tic:
    if (lst_ele[1] not in lst_tech_sp500_tic_sing) or ((lst_ele[2]!="None") and (lst_ele[2] not in lst_tech_sp500_tic_sing)): 
        print(lst_ele, lst_gpt_gen_new_old_tic.index(lst_ele))
        
print("no ono")

['Alphabet Inc.', 'GOOGL/GOOG', 'None'] 3
['Veralto Corp', 'VRTX', 'None'] 64
['Fox Corp', 'FOXA/FOXA', 'None'] 81
no ono


In [50]:
("GOOGL" in lst_tech_sp500_tic_sing) and ("GOOG" in lst_tech_sp500_tic_sing)

True

In [51]:
"FOXA" in lst_tech_sp500_tic_sing

True

The only one need to be fixed is ['Veralto Corp', 'VRTX', 'None'] 64, a short google shows: "VLTO" is the stock ticker symbol for Veralto Corporation, a company focused on water analytics and treatment solutions, while "VRTX" is the stock ticker symbol for Vertex Pharmaceuticals Incorporated, a biotechnology company specializing in medicines for serious diseases. So i think chat gpt was wrong on this one, we will not edit anything. 

We check for the weird tickers in lst_tech_sp500_tic_sing again: 

In [56]:
reg_allow=re.compile("^[A-Za-z0-9]*$")
lst_tech_sp500_tic_sing_weird=[tic for tic in lst_tech_sp500_tic_sing if not bool(reg_allow.fullmatch(tic))]

In [57]:
lst_tech_sp500_tic_sing_weird

['T-PA', 'T-PC', 'HPE-PC', 'PBI-PB']

We check that the company tickers in the lst_tech_sp500_tic_sing_weird are all in lst_tech_sp500_tic_sing: 

In [58]:
for tic in lst_tech_sp500_tic_sing_weird: 
    com_tic=re.findall("([A-Za-z0-9]*)\-",tic)[0]
    if com_tic not in lst_tech_sp500_tic_sing: 
        print("ono:",tic,com_tic)
print("no ono")

no ono


<>:2: SyntaxWarning: invalid escape sequence '\-'
<>:2: SyntaxWarning: invalid escape sequence '\-'
/tmp/ipykernel_13098/278471754.py:2: SyntaxWarning: invalid escape sequence '\-'
  com_tic=re.findall("([A-Za-z0-9]*)\-",tic)[0]


Import his_weight_fix_tic_in_process.csv, as a reminder, according to https://www.kaggle.com/datasets/akshaytolwani/s-and-p-500-constituents-with-weights-2009-2024?select=sp500_historical.csv, daily data is only available starting 2012-04-30, so we will need to eliminate the data before that point. 

In [59]:
df_his_weight=pd.read_csv("../../processed/his_weight_fix_tic_in_process.csv")

In [60]:
df_his_weight

date   rank ticker                      name  \
0        2009-01-30    1.0    XOM          EXXON MOBIL CORP   
1        2009-01-30    2.0     PG          PROCTER & GAMBLE   
2        2009-01-30    3.0    JNJ         JOHNSON & JOHNSON   
3        2009-01-30    4.0      T                  AT&T INC   
4        2009-01-30    5.0    CVX              CHEVRON CORP   
...             ...    ...    ...                       ...   
1527403  2024-10-30  499.0    BEN    FRANKLIN RESOURCES INC   
1527404  2024-10-30  500.0   PARA  PARAMOUNT GLOBAL CLASS B   
1527405  2024-10-30  501.0    FOX          FOX CORP CLASS B   
1527406  2024-10-30  502.0   AMTM      AMENTUM HOLDINGS INC   
1527407  2024-10-30  503.0    NWS         NEWS CORP CLASS B   

                             sector    weight  
0                            Energy  5.403875  
1                  Consumer Staples  2.260313  
2                       Health Care  2.223419  
3        Telecommunication Services  2.015346  
4                            Energy  1.990291  
...                             ...       ...  
1527403                  Financials  0.013080  
1527404               Communication  0.012860  
1527405               Communication  0.010430  
1527406                 Industrials  0.007630  
1527407               Communication  0.006670  

[1527408 rows x 6 columns]

In [71]:
df_his_weight_12_24=df_his_weight[df_his_weight.date >= "2012-04-30"]

In [72]:
df_his_weight_12_24

date   rank ticker                                name  \
20005    2012-04-30    1.0  AAPL*                           APPLE INC   
20006    2012-04-30    2.0    XOM                    EXXON MOBIL CORP   
20007    2012-04-30    3.0    IBM  INTERNATIONAL BUSINESS MACHINES CO   
20008    2012-04-30    4.0   MSFT                      MICROSOFT CORP   
20009    2012-04-30    5.0    CVX                        CHEVRON CORP   
...             ...    ...    ...                                 ...   
1527403  2024-10-30  499.0    BEN              FRANKLIN RESOURCES INC   
1527404  2024-10-30  500.0   PARA            PARAMOUNT GLOBAL CLASS B   
1527405  2024-10-30  501.0    FOX                    FOX CORP CLASS B   
1527406  2024-10-30  502.0   AMTM                AMENTUM HOLDINGS INC   
1527407  2024-10-30  503.0    NWS                   NEWS CORP CLASS B   

                         sector    weight  
20005    Information Technology  4.306649  
20006                    Energy  3.217278  
20007    Information Technology  1.896926  
20008    Information Technology  1.890466  
20009                    Energy  1.665514  
...                         ...       ...  
1527403              Financials  0.013080  
1527404           Communication  0.012860  
1527405           Communication  0.010430  
1527406             Industrials  0.007630  
1527407           Communication  0.006670  

[1507403 rows x 6 columns]

Back up the date

In [ ]:
df_his_weight_12_24.to_csv("../../processed/his_weight_12_24_in_progress.csv", index=False)

In [76]:
df_his_weight_12_24=pd.read_csv("../../processed/his_weight_12_24_in_progress.csv")

In [77]:
df_his_weight_12_24

date   rank ticker                                name  \
0        2012-04-30    1.0  AAPL*                           APPLE INC   
1        2012-04-30    2.0    XOM                    EXXON MOBIL CORP   
2        2012-04-30    3.0    IBM  INTERNATIONAL BUSINESS MACHINES CO   
3        2012-04-30    4.0   MSFT                      MICROSOFT CORP   
4        2012-04-30    5.0    CVX                        CHEVRON CORP   
...             ...    ...    ...                                 ...   
1507398  2024-10-30  499.0    BEN              FRANKLIN RESOURCES INC   
1507399  2024-10-30  500.0   PARA            PARAMOUNT GLOBAL CLASS B   
1507400  2024-10-30  501.0    FOX                    FOX CORP CLASS B   
1507401  2024-10-30  502.0   AMTM                AMENTUM HOLDINGS INC   
1507402  2024-10-30  503.0    NWS                   NEWS CORP CLASS B   

                         sector    weight  
0        Information Technology  4.306649  
1                        Energy  3.217278  
2        Information Technology  1.896926  
3        Information Technology  1.890466  
4                        Energy  1.665514  
...                         ...       ...  
1507398              Financials  0.013080  
1507399           Communication  0.012860  
1507400           Communication  0.010430  
1507401             Industrials  0.007630  
1507402           Communication  0.006670  

[1507403 rows x 6 columns]

In [82]:
len(list(df_his_weight_12_24.date.unique()))

2994

Going to check if a company ticker in df_his_weight_12_24 can be found in lst_tech_sp500_tic_sing

We review the tickers of df_his_weight_12_24: 

In [78]:
lst_his_weight_tic=list(df_his_weight_12_24.ticker.unique())

In [79]:
lst_his_weight_tic

['AAPL*',
 'XOM',
 'IBM',
 'MSFT',
 'CVX',
 'GEC',
 'T',
 'JNJ',
 'WFC*',
 'PG',
 'PFE',
 'KO',
 'JPM',
 'PM',
 'GOOGL',
 'INCO',
 'BRKB',
 'MRK',
 'ORCL',
 'VZ',
 'CSCO',
 'QCI',
 'PEP',
 'WMT',
 'MCD',
 'SLBA',
 'ABT',
 'C',
 'COP',
 'BAC',
 'AMZN',
 'CMCSA',
 'HD',
 'DIS',
 'UPS',
 'UTX',
 'OXY',
 'KFT',
 'CAT',
 'MO',
 'MMM',
 'USB',
 'VISA',
 'AEC1',
 'UNH',
 'CVS',
 'EMC',
 'BA',
 'GS',
 'BMY',
 'AMGN',
 'UNP',
 'DD',
 'HPQ',
 'MA',
 'SPG',
 'CL',
 'ALD',
 'EBAY',
 '4XS',
 'SBUX',
 'F',
 'LLY',
 'NCRA',
 'ACN',
 'NKE',
 'MON',
 'DOW',
 'SO',
 'MDT',
 'GILD',
 'LOW',
 'TGT',
 'EMR',
 'COST',
 'MWZ',
 'PCLN',
 'APA',
 'TXN',
 'AAZ',
 'TWX*',
 'FCX*',
 'PNC',
 'PX',
 'YUM',
 'DTV',
 'EXC',
 'DE',
 'NOV',
 'CELG',
 'BIIB',
 'HAL',
 'TJX',
 'BAX',
 'DHR',
 'KMB',
 'COF',
 'WAG',
 'D',
 'EOG',
 'AG4',
 'DUK',
 'BK',
 'PRU*',
 'DVN',
 'ITW',
 'FDX',
 'ADP',
 'NEE',
 'COV',
 'MS',
 'TYC',
 'AMT',
 'CB',
 'PCP',
 'PA9',
 'T3W1',
 'GIS',
 'VIAB',
 'DELL',
 'LOM',
 'GD',
 'NSC',
 'LUMN',
 '

We review the weird tickers of df_his_weight

In [80]:
lst_his_weight_tic_wrd=[tic for tic in lst_his_weight_tic if not bool(reg_allow.fullmatch(tic))]

In [81]:
lst_his_weight_tic_wrd

['AAPL*',
 'WFC*',
 'TWX*',
 'FCX*',
 'PRU*',
 'ADM*',
 'AIG*',
 'MOS*',
 'KMI WS',
 'S*',
 'FB*',
 'GOOG*',
 'BF/B',
 'BF.B',
 'TLN WI',
 'GCI.WI',
 'BLD WI',
 'CC WI',
 'CCP W',
 'SYF-W',
 'UAC/C',
 'HWM WI',
 'ARNC WI',
 'RTX US',
 'OXY WS WI',
 'OXY WS',
 'VTRS US',
 'MBC WI']

In [85]:
def tic_is_in(str_tic): 
    if str_tic in lst_his_weight_tic_wrd: 
        com_tic=re.findall("([A-Za-z0-9]*)[^A-Za-z0-9]",str_tic)[0]
        return com_tic in lst_tech_sp500_tic_sing
    else: 
        return str_tic in lst_tech_sp500_tic_sing 

In [87]:
tic_is_in("AAPL")

True

In [88]:
lst_weight_tic_in_tech_500=[str_tic for str_tic in lst_his_weight_tic if tic_is_in(str_tic)]

In [89]:
lst_weight_tic_in_tech_500

['AAPL*',
 'IBM',
 'MSFT',
 'T',
 'GOOGL',
 'ORCL',
 'VZ',
 'CSCO',
 'CMCSA',
 'CAT',
 'HPQ',
 'MDT',
 'TXN',
 'DE',
 'NOV',
 'DHR',
 'ITW',
 'ADP',
 'DELL',
 'LUMN',
 'CTSH',
 'CMI',
 'JCI',
 'CRM',
 'TMO',
 'BRCM',
 'INTU',
 'ADBE',
 'ETN',
 'AMAT',
 'NTAP',
 'SLE',
 'STI',
 'TT',
 'TDC',
 'ADI',
 'DOV',
 'ALTR',
 'JNPR',
 'FTI',
 'ROK',
 'FFIV',
 'XRX',
 'ROP',
 'APH',
 'WDC',
 'ADSK',
 'NVDA',
 'S',
 'MCHP',
 'VRSN',
 'MU',
 'FLS',
 'XYL',
 'EA',
 'AMD',
 'HAS',
 'TRIP',
 'PBI',
 'LDOS',
 'TGNA',
 'TER',
 'DV',
 'FSLR',
 'LRCX',
 'STX',
 'PNR',
 'TMUS',
 'FOXA',
 'S*',
 'AME',
 'FB*',
 'AAPL',
 'FB',
 'GOOG*',
 'INTC',
 'GOOG',
 'AVGO',
 'DNOW',
 'KEYS',
 'HCP',
 'IR',
 'SWKS',
 'QRVO',
 'FOX',
 'VRSK',
 'HPE',
 'AYI',
 'FTV',
 'CHTR',
 'DXC',
 'SNPS',
 'ANSS',
 'CDNS',
 'IPGP',
 'TTWO',
 'ANET',
 'FTNT',
 'JKHY',
 'IEX',
 'BKR',
 'NOW',
 'ZBRA',
 'PAYC',
 'CARR',
 'TYL',
 'ENPH',
 'TRMB',
 'MPWR',
 'NXPI',
 'PTC',
 'MTCH',
 'EPAM',
 'FDS',
 'SEDG',
 'NDSN',
 'PARA',
 'WBD',
 'META

In [90]:
df_his_weight_tech_500_12_24=df_his_weight_12_24[df_his_weight_12_24.ticker.isin(lst_weight_tic_in_tech_500)]

In [91]:
df_his_weight_tech_500_12_24

date   rank ticker                                name  \
0        2012-04-30    1.0  AAPL*                           APPLE INC   
2        2012-04-30    3.0    IBM  INTERNATIONAL BUSINESS MACHINES CO   
3        2012-04-30    4.0   MSFT                      MICROSOFT CORP   
6        2012-04-30    7.0      T                            AT&T INC   
14       2012-04-30   15.0  GOOGL                  GOOGLE INC CLASS A   
...             ...    ...    ...                                 ...   
1507384  2024-10-30  485.0    HAS                          HASBRO INC   
1507388  2024-10-30  489.0   PAYC                 PAYCOM SOFTWARE INC   
1507395  2024-10-30  496.0   QRVO                           QORVO INC   
1507399  2024-10-30  500.0   PARA            PARAMOUNT GLOBAL CLASS B   
1507400  2024-10-30  501.0    FOX                    FOX CORP CLASS B   

                             sector    weight  
0            Information Technology  4.306649  
2            Information Technology  1.896926  
3            Information Technology  1.890466  
6        Telecommunication Services  1.542570  
14           Information Technology  1.213489  
...                             ...       ...  
1507384      Consumer Discretionary  0.017720  
1507388                 Industrials  0.017020  
1507395      Information Technology  0.014060  
1507399               Communication  0.012860  
1507400               Communication  0.010430  

[237263 rows x 6 columns]

Back up the data

In [92]:
df_his_weight_tech_500_12_24.to_csv("../../processed/his_weight_tech_500_12_24_in_progress.csv",index=False)

In [93]:
df_his_weight_tech_500_12_24=pd.read_csv("../../processed/his_weight_tech_500_12_24_in_progress.csv")

In [94]:
df_his_weight_tech_500_12_24

date   rank ticker                                name  \
0       2012-04-30    1.0  AAPL*                           APPLE INC   
1       2012-04-30    3.0    IBM  INTERNATIONAL BUSINESS MACHINES CO   
2       2012-04-30    4.0   MSFT                      MICROSOFT CORP   
3       2012-04-30    7.0      T                            AT&T INC   
4       2012-04-30   15.0  GOOGL                  GOOGLE INC CLASS A   
...            ...    ...    ...                                 ...   
237258  2024-10-30  485.0    HAS                          HASBRO INC   
237259  2024-10-30  489.0   PAYC                 PAYCOM SOFTWARE INC   
237260  2024-10-30  496.0   QRVO                           QORVO INC   
237261  2024-10-30  500.0   PARA            PARAMOUNT GLOBAL CLASS B   
237262  2024-10-30  501.0    FOX                    FOX CORP CLASS B   

                            sector    weight  
0           Information Technology  4.306649  
1           Information Technology  1.896926  
2           Information Technology  1.890466  
3       Telecommunication Services  1.542570  
4           Information Technology  1.213489  
...                            ...       ...  
237258      Consumer Discretionary  0.017720  
237259                 Industrials  0.017020  
237260      Information Technology  0.014060  
237261               Communication  0.012860  
237262               Communication  0.010430  

[237263 rows x 6 columns]

In [95]:
len(df_his_weight_tech_500_12_24.date.unique())

2994

Turn each line of lst_tech_sp500_tic into a set

In [96]:
lst_tech_sp500_tic_set=[set(str_tics.split(",")) for str_tics in lst_tech_sp500_tic]

In [97]:
lst_tech_sp500_tic_set

[{'AAPL'},
 {'MSFT'},
 {'NVDA'},
 {'GOOG', 'GOOGL'},
 {'FB', 'META'},
 {'AVGO', 'BRCM'},
 {'ORCL'},
 {'CRM'},
 {'ADBE'},
 {'AMD'},
 {'TMO'},
 {'TMUS'},
 {'CSCO'},
 {'DHR'},
 {'INTU'},
 {'TXN'},
 {'IBM'},
 {'AMAT'},
 {'NOW'},
 {'VZ'},
 {'CAT'},
 {'CCZ', 'CMCSA'},
 {'T', 'T-PA', 'T-PC', 'TBB', 'TBC'},
 {'MU'},
 {'ETN'},
 {'LRCX'},
 {'ANET'},
 {'ADI'},
 {'PANW'},
 {'MDT'},
 {'ADP'},
 {'DE'},
 {'INTC'},
 {'SNPS'},
 {'APH'},
 {'DELL'},
 {'CDNS'},
 {'TT'},
 {'PLTR'},
 {'ITW'},
 {'NXPI'},
 {'CRWD'},
 {'CARR'},
 {'ROP'},
 {'FTNT'},
 {'ADSK'},
 {'CHTR'},
 {'JCI'},
 {'MPWR'},
 {'MCHP'},
 {'CMI'},
 {'EA'},
 {'VRSK'},
 {'AME'},
 {'CTSH'},
 {'IR'},
 {'SMCI'},
 {'HPQ'},
 {'BKR'},
 {'ON'},
 {'XYL'},
 {'ROK'},
 {'ANSS'},
 {'NTAP'},
 {'VLTO'},
 {'TTWO'},
 {'TYL'},
 {'DOV'},
 {'HPE', 'HPE-PC'},
 {'FTV'},
 {'FSLR'},
 {'KEYS'},
 {'GDDY'},
 {'TER'},
 {'STX'},
 {'WDC'},
 {'PTC'},
 {'HUBB'},
 {'LDOS'},
 {'WBD'},
 {'ZBRA'},
 {'FOX', 'FOXA'},
 {'VRSN'},
 {'SWKS'},
 {'ENPH'},
 {'FDS'},
 {'GEN'},
 {'IEX'},
 {'PN

Add in the weird tickers from lst_weight_tic_in_tech_500 

In [98]:
for set_ele in lst_tech_sp500_tic_set: 
    for str_tic in lst_his_weight_tic_wrd: 
        if (re.findall("([A-Za-z0-9]*)[^A-Za-z0-9]",str_tic)[0] in set_ele) and (str_tic not in set_ele): 
            set_ele.add(str_tic)

In [101]:
lst_tech_sp500_tic_set

[{'AAPL', 'AAPL*'},
 {'MSFT'},
 {'NVDA'},
 {'GOOG', 'GOOG*', 'GOOGL'},
 {'FB', 'FB*', 'META'},
 {'AVGO', 'BRCM'},
 {'ORCL'},
 {'CRM'},
 {'ADBE'},
 {'AMD'},
 {'TMO'},
 {'TMUS'},
 {'CSCO'},
 {'DHR'},
 {'INTU'},
 {'TXN'},
 {'IBM'},
 {'AMAT'},
 {'NOW'},
 {'VZ'},
 {'CAT'},
 {'CCZ', 'CMCSA'},
 {'T', 'T-PA', 'T-PC', 'TBB', 'TBC'},
 {'MU'},
 {'ETN'},
 {'LRCX'},
 {'ANET'},
 {'ADI'},
 {'PANW'},
 {'MDT'},
 {'ADP'},
 {'DE'},
 {'INTC'},
 {'SNPS'},
 {'APH'},
 {'DELL'},
 {'CDNS'},
 {'TT'},
 {'PLTR'},
 {'ITW'},
 {'NXPI'},
 {'CRWD'},
 {'CARR'},
 {'ROP'},
 {'FTNT'},
 {'ADSK'},
 {'CHTR'},
 {'JCI'},
 {'MPWR'},
 {'MCHP'},
 {'CMI'},
 {'EA'},
 {'VRSK'},
 {'AME'},
 {'CTSH'},
 {'IR'},
 {'SMCI'},
 {'HPQ'},
 {'BKR'},
 {'ON'},
 {'XYL'},
 {'ROK'},
 {'ANSS'},
 {'NTAP'},
 {'VLTO'},
 {'TTWO'},
 {'TYL'},
 {'DOV'},
 {'HPE', 'HPE-PC'},
 {'FTV'},
 {'FSLR'},
 {'KEYS'},
 {'GDDY'},
 {'TER'},
 {'STX'},
 {'WDC'},
 {'PTC'},
 {'HUBB'},
 {'LDOS'},
 {'WBD'},
 {'ZBRA'},
 {'FOX', 'FOXA'},
 {'VRSN'},
 {'SWKS'},
 {'ENPH'},
 {'FDS'},


In [125]:
df_his_weight_tech_500_12_24[df_his_weight_tech_500_12_24.ticker.isin(lst_tech_sp500_tic_set[4])]

date  rank ticker                        name  \
26567   2013-12-23  36.0    FB*       FACEBOOK CLASS A  INC   
26630   2013-12-24  35.0    FB*       FACEBOOK CLASS A  INC   
26693   2013-12-26  37.0    FB*       FACEBOOK CLASS A  INC   
26756   2013-12-27  37.0    FB*       FACEBOOK CLASS A  INC   
26819   2013-12-30  38.0    FB*       FACEBOOK CLASS A  INC   
...            ...   ...    ...                         ...   
236751  2024-10-24   5.0   META  META PLATFORMS INC CLASS A   
236854  2024-10-25   5.0   META  META PLATFORMS INC CLASS A   
236957  2024-10-28   5.0   META  META PLATFORMS INC CLASS A   
237060  2024-10-29   5.0   META  META PLATFORMS INC CLASS A   
237163  2024-10-30   5.0   META  META PLATFORMS INC CLASS A   

                        sector    weight  
26567   Information Technology  0.579483  
26630   Information Technology  0.579631  
26693   Information Technology  0.574605  
26756   Information Technology  0.551890  
26819   Information Technology  0.536758  
...                        ...       ...  
236751           Communication  2.519030  
236854           Communication  2.544040  
236957           Communication  2.559020  
237060           Communication  2.621700  
237163           Communication  2.623810  

[2583 rows x 6 columns]

Create a list determining if a company in df_tech_sp500 is stable or not. There are 2994 distinct dates in df_his_weight_tech_500_12_24, so if there are at least that much rows for a company, it should be stable. There might be some missing data, so we will leave a little room and change 2994 to 2990. 

In [121]:
df_his_weight_tech_500_12_24[df_his_weight_tech_500_12_24.name.isin(lst_tech_sp500_tic_set[1])]

Empty DataFrame
Columns: [date, rank, ticker, name, sector, weight]
Index: []

In [138]:
lst_stable=[]

for ele_set in lst_tech_sp500_tic_set: 
    lst_stable.append(len(df_his_weight_tech_500_12_24[df_his_weight_tech_500_12_24.ticker.isin(ele_set)])>=2990)

In [139]:
lst_stable

[True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 False,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 True,
 False,
 True,
 True,
 True,
 False,
 False,
 True,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 True,
 False,
 True,
 False,
 True,
 False,
 True,
 True,
 True,
 False,
 False,
 True,
 False,
 False,
 True,
 False,
 False,
 True,
 True,
 False,
 True,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 True,
 False,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False]

In [140]:
lst_stable_count=[]

for ele_set in lst_tech_sp500_tic_set: 
    lst_stable_count.append(len(df_his_weight_tech_500_12_24[df_his_weight_tech_500_12_24.ticker.isin(ele_set)]))

In [141]:
lst_stable_count

[2994,
 2993,
 2994,
 5501,
 2583,
 3415,
 2993,
 2994,
 2994,
 2187,
 2994,
 1333,
 2994,
 2994,
 2994,
 2994,
 2994,
 2994,
 1248,
 2994,
 2994,
 2994,
 2994,
 2994,
 2963,
 2969,
 1552,
 2994,
 346,
 2993,
 2993,
 2994,
 2496,
 1839,
 2994,
 404,
 1790,
 1811,
 29,
 2994,
 915,
 92,
 1156,
 2994,
 1522,
 2993,
 1920,
 2994,
 940,
 2994,
 2993,
 2993,
 2149,
 2646,
 2993,
 2363,
 159,
 2994,
 1271,
 596,
 2994,
 2994,
 1839,
 2994,
 273,
 1665,
 1103,
 2994,
 2131,
 1965,
 1626,
 1508,
 92,
 1451,
 2951,
 2994,
 894,
 262,
 1660,
 643,
 1227,
 4870,
 2994,
 2295,
 965,
 721,
 498,
 1312,
 2891,
 682,
 956,
 2994,
 1499,
 2054,
 725,
 2994,
 0,
 2231,
 2993,
 1204,
 786,
 189,
 679,
 0,
 902,
 405,
 298,
 2208,
 1105,
 2078,
 1261,
 1566,
 104,
 1077,
 1155,
 1010,
 1767,
 501,
 5,
 1155,
 2079,
 0,
 1758,
 42]

We append above into the df_tech_sp500. 

In [142]:
df_tech_sp500["stable"]=lst_stable

In [144]:
df_tech_sp500.rename(columns={"stable":"stable 12 to 24"})

cik                           name entityType     sic  \
0     320193                     Apple Inc.  operating  3571.0   
1     789019                 MICROSOFT CORP  operating  7372.0   
2    1045810                    NVIDIA CORP  operating  3674.0   
3    1652044                  Alphabet Inc.  operating  7370.0   
4    1326801           Meta Platforms, Inc.  operating  7370.0   
..       ...                            ...        ...     ...   
119    78814          PITNEY BOWES INC /DE/  operating  3579.0   
120  1770450            Xerox Holdings Corp  operating  3577.0   
121    61986               MANITOWOC CO INC  operating  3531.0   
122  1881551       Solidion Technology Inc.  operating  3690.0   
123  1621672  Super League Enterprise, Inc.  operating  7373.0   

                                        sicDescription          ownerOrg  \
0                                 Electronic Computers     06 Technology   
1                        Services-Prepackaged Software     06 Technology   
2                     Semiconductors & Related Devices  04 Manufacturing   
3    Services-Computer Programming, Data Processing...     06 Technology   
4    Services-Computer Programming, Data Processing...     06 Technology   
..                                                 ...               ...   
119                               Office Machines, NEC     06 Technology   
120                 Computer Peripheral Equipment, NEC     06 Technology   
121                     Construction Machinery & Equip     06 Technology   
122  Miscellaneous Electrical Machinery, Equipment ...  04 Manufacturing   
123        Services-Computer Integrated Systems Design     06 Technology   

     insiderTransactionForOwnerExists  insiderTransactionForIssuerExists  \
0                                   0                                  1   
1                                   1                                  1   
2                                   1                                  1   
3                                   1                                  1   
4                                   0                                  1   
..                                ...                                ...   
119                                 0                                  1   
120                                 0                                  1   
121                                 1                                  1   
122                                 0                                  1   
123                                 0                                  1   

        tickers      exchanges  ...  mailing_zipCode  \
0          AAPL         Nasdaq  ...            95014   
1          MSFT         Nasdaq  ...       98052-6399   
2          NVDA         Nasdaq  ...            95051   
3    GOOGL,GOOG  Nasdaq,Nasdaq  ...            94043   
4       META,FB         Nasdaq  ...            94025   
..          ...            ...  ...              ...   
119  PBI,PBI-PB      NYSE,NYSE  ...       06926-0700   
120         XRX         Nasdaq  ...            06851   
121         MTW           NYSE  ...            53224   
122         STI         Nasdaq  ...            75240   
123         SLE         Nasdaq  ...            90404   

     mailing_stateOrCountryDescription            business_street1  \
0                                   CA          ONE APPLE PARK WAY   
1                                   WA           ONE MICROSOFT WAY   
2                                   CA   2788 SAN TOMAS EXPRESSWAY   
3                                   CA   1600 AMPHITHEATRE PARKWAY   
4                                   CA                  1 META WAY   
..                                 ...                         ...   
119                                 CT            PITNEY BOWES INC   
120                                 CT               201 MERRITT 7   
121                                 WI       11270 WEST PARK PLACE   
122              

Write data

In [145]:
df_tech_sp500.to_csv("../../processed/tech_sp500_stable.csv", index=False)

We might want to make some further (manual) decisions. 